pymatgen/core/trajectory.py

```diff
@@ -537,20 +539,36 @@ def from_file(cls, filename: str | Path, constant_lattice: bool = True, **kwargs
             A trajectory from the file.
         """
         fname = Path(filename).expanduser().resolve().name
+        is_mol = False
 
         if fnmatch(fname, "*XDATCAR*"):
             structures = Xdatcar(filename).structures
         elif fnmatch(fname, "vasprun*.xml*"):
             structures = Vasprun(filename).structures
+        elif fnmatch(fname, "*.traj"):
+            ase_traj = ASE_Trajectory(fname)
+            pbc = ase_traj[0].pbc
+            if any(pbc):
+                structures = [AseAtomsAdaptor.get_structure(atoms) for atoms in ase_traj]
+            else:
+                molecules = [AseAtomsAdaptor.get_molecule(atoms) for atoms in ase_traj]
+                is_mol = True
+
         else:
-            supported = ("XDATCAR", "vasprun.xml")
+            supported = ("XDATCAR", "vasprun.xml", "*.traj")
             raise ValueError(f"Expect file to be one of {supported}; got {filename}.")
 
-        return cls.from_structures(
-            structures,
-            constant_lattice=constant_lattice,
-            **kwargs,
-        )
+        if not is_mol:
```

Comment:
You should do `if is_mol` (return from_molecules) and then `else` (return from_structures). Makes more sense to start with a positive logic statement rather than a negative.